protect_archiver/client.py

```diff
@@ -0,0 +1,372 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+""" Tool to download footage within a given time range from a local UniFi Protect system """
+
+import logging
+import time
+
+from dataclasses import dataclass
+from datetime import datetime
+from os import path, makedirs
+from typing import List
+
+import requests
+
+from .utils import calculate_intervals
+
+
+@dataclass
+class Camera:
+    id: str
+    name: str
+    recording_start: datetime
+
+
+class DownloadFailed(Exception):
+    pass
+
+
+class ProtectError(Exception):
+    def __init__(self, code):
+        self.code = code
+
+
+class ProtectClient(object):
+    def __init__(
+        self,
+        address: str = "unifi",
+        port: int = 7443,
+        username: str = "ubnt",
+        password: str = None,
+        verify_ssl: bool = False,
+        ignore_failed_downloads: bool = False,
+        download_wait: int = 0,
+        use_subfolders: bool = False,
+        skip_existing_files: bool = False,
+        destination_path: str = "./",
+        touch_files: bool = False,
+        download_timeout: int = 60,
+        max_downloads_with_key: int = 3,
+        max_downloads_with_auth: int = 10,
+    ):
+        self.address = address
+        self.port = port
+        self.username = username
+        self.password = password
+        self.verify_ssl = verify_ssl
+
+        self.ignore_failed_downloads = ignore_failed_downloads
+        self.download_wait = download_wait
+        self.download_timeout = download_timeout
+        self.use_subfolders = use_subfolders
+        self.skip_existing_files = skip_existing_files
+        self.touch_files = touch_files
+
+        self.destination_path = path.abspath(destination_path)
+
+        self.files_downloaded = 0
+        self.files_skipped = 0
+        self.max_retries = 3
+
+        self.max_downloads_with_auth = max_downloads_with_auth
```

Comment:
it'd be nice to document why you might need to change these, and if we dont think you do we could just make them constants